File: dsa_study/src/dsa_study/array/MoveZeros.cs
namespace dsa_study.array;

public class MoveZeros
{
    // Método principal para executar demonstrações de funcionalidades da classe.
    // static void Main()
    // {
    //     MoveZeros moveZeros = new();
    //     moveZeros.arrayDemo();
    // }

    // Método para demonstrar a funcionalidade de mover zeros para o fim de um array.
    public void arrayDemo()
    {
        int[] array = { 8, 1, 0, 2, 1, 0, 3 };
        printArray(array); // Imprime o array antes da operação
        moveZeros(array, array.Length); // Executa o método para mover zeros
        printArray(array); // Imprime o array após mover os zeros
    }

    // Imprime todos os elementos de um array, separados por espaço.
    public void printArray(int[] array)
    {
        int numeroDeElementos = array.Length;

        for (int i = 0; i < numeroDeElementos; i++)
        {
            Console.Write(array[i] + " "); // Imprime cada elemento seguido de um espaço
        }
        Console.WriteLine(); // Finaliza com uma nova linha
    }

    // Método para mover todos os zeros de um array para o seu final, mantendo a ordem relativa dos outros elementos.
    public void moveZeros(int[] array, int numeroDeElementos)
    {
        int j = 0; // Inicializa o ponteiro 'j' para a posição do primeiro elemento não-zero
        for (int i = 0; i < numeroDeElementos; i++)
        {
            if (array[i] != 0)
            {
                // Se o elemento atual não for zero e 'j' for menor que 'i', troca os elementos
                if (i != j)
                {
                    int temp = array[i];
                    array[i] = array[j];
                    array[j] = temp;
                }
                j++; // Incrementa 'j' para apontar para o próximo elemento potencialmente zero
            }
        }
    }
}

File: go/dsa/array/MoveZeros/MoveZeros.go
package movezeros

import "fmt"

// moveZeros move todos os zeros no array para o final, mantendo a ordem relativa dos elementos não-zero.
func moveZeros(array []int) {
	j := 0 // Inicializa o ponteiro 'j' para rastrear a posição do próximo elemento não-zero
	for i := 0; i < len(array); i++ { // Percorre o array
		if array[i] != 0 { // Se o elemento atual não for zero
			if i != j { // Se o índice atual 'i' não for igual ao índice 'j'
				array[i], array[j] = array[j], array[i] // Troca os elementos nas posições 'i' e 'j'
			}
			j++ // Incrementa o ponteiro 'j' para apontar para o próximo elemento potencialmente zero
		}
	}
}

// printArray imprime todos os elementos de um array, separados por um espaço.
func printArray(array []int) {
	for _, v := range array { // Percorre cada elemento do array
		fmt.Printf("%d ", v) // Imprime o elemento seguido de um espaço
	}
	fmt.Println() // Finaliza com uma nova linha
}

// arrayDemo demonstra a funcionalidade de moveZeros e printArray.
func arrayDemo() {
	array := []int{8, 1, 0, 2, 1, 0, 3} // Inicializa um array com elementos incluindo zeros
	printArray(array) // Imprime o array original
	moveZeros(array) // Move todos os zeros para o final do array
	printArray(array) // Imprime o array após mover os zeros
}

